src/jit/function-builder.cc

```diff
@@ -832,78 +771,74 @@ bool FunctionBuilder::Emit(TR::BytecodeBuilder* b,
       break;
     }
 
-    case Opcode::Call: {
-      auto th_addr = b->ConstAddress(thread_);
-      auto fn = cast<interp::DefinedFunc>(thread_->env_->GetFunc(interp::ReadU32(&pc)));
-      auto current_pc = b->Const(pc);
+    case Opcode::Call:
+    case Opcode::InterpCallHost: {
+      auto fn_ind = interp::ReadU32(&pc);
+      auto fn = thread_->env_->GetFunc(fn_ind);
 
       auto* sig = thread_->env()->GetFuncSignature(fn->sig_index);
 
+      auto* thread = b->Load("thread");
+      auto* call_stack = b->LoadIndirect("ThreadInfo", "call_stack", thread);
+
       MoveToPhysStack(b, pc, &stack, sig->param_types.size());
 
-      b->Store("result",
-      b->      Call("CallHelper", 3, th_addr, b->ConstAddress(fn), current_pc));
+      auto* fn_thunk = b->LoadAt(typeDictionary()->pAddress,
+                       b->       IndexAt(typeDictionary()->pAddress,
+                       b->               LoadIndirect("ThreadInfo", "jit_fn_table", thread),
+                       b->               ConstInt32(fn_ind)));
+
+      EmitTrapIf(b,
+      b->        EqualTo(call_stack, b->LoadIndirect("ThreadInfo", "call_stack_max", thread)),
+      b->        Const(static_cast<Result_t>(interp::Result::TrapCallStackExhausted)),
+                 pc);
+
+      b->StoreIndirect("ThreadInfo", "call_stack", thread, b->IndexAt(typeDictionary()->pAddress, call_stack, b->ConstInt32(1)));
+      b->StoreIndirect("CallFrame", "pc", call_stack, b->ConstInt32(pc - thread_->GetIstream()));
+      b->StoreIndirect("CallFrame", "is_jit", call_stack, b->ConstInt8(1));
+      b->StoreIndirect("CallFrame", "is_jit_compiling", call_stack, b->ConstInt8(0));
```

Comment:
Are we sure that these booleans are 8 bits wide?

Reply:
Good catch. No, nothing guarantees that these booleans will be 8 bits wide... They were in my testing so I temporarily assumed they were when I was writing the initial prototype, but I guess I forgot to fix that before submitting the changes... :sweat_smile: